tools/workflow/cleaner/iam/clean.go

```diff
@@ -16,73 +16,157 @@
 package iam
 
 import (
+	"fmt"
 	"log"
+	"os"
 	"strings"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/iam"
 )
 
 const Type = "iam"
 
-var roleNamePrefixes = []string{"terraform", "aoc-eks-assume-role", "fargate-profile-role"}
+var (
+	logger             = log.New(os.Stdout, fmt.Sprintf("[%s] ", Type), log.Ldate)
+	roleNamePrefixes   = []string{"terraform", "aoc-eks-assume-role", "fargate-profile-role"}
+	policyNamePrefixes = []string{"terraform"}
+)
 
-func Clean(sess *session.Session, keepDuration time.Duration) {
+func Clean(sess *session.Session, keepDuration time.Duration) error {
+	logger.Printf("Begin to clean IAM roles")
 	client := iam.New(sess)
+	expirationDate := time.Now().UTC().Add(keepDuration)
+	if err := deleteRoles(client, expirationDate); err != nil {
+		return err
+	}
+	if err := deletePolicies(client, expirationDate); err != nil {
+		return err
+	}
+	return nil
+}
 
-	var rolesMarker *string
-
+func deleteRoles(client *iam.IAM, expirationDate time.Time) error {
+	var marker *string
 	for {
-		lri := &iam.ListRolesInput{Marker: rolesMarker}
+		lri := &iam.ListRolesInput{Marker: marker}
 		lro, err := client.ListRoles(lri)
-
 		if err != nil {
-			log.Fatalf("Failed to get roles because of %v", err)
+			return err
 		}
-
 		for _, role := range lro.Roles {
-			var hasPrefix bool
-			for _, prefix := range roleNamePrefixes {
-				if hasPrefix = strings.HasPrefix(*role.RoleName, prefix); hasPrefix {
-					break
+			if hasPrefix(roleNamePrefixes, *role.RoleName) && expirationDate.After(*role.CreateDate) &&
+				(role.RoleLastUsed == nil || expirationDate.After(*role.RoleLastUsed.LastUsedDate)) {
+				logger.Printf("Trying to delete IAM role %s created on %v", *role.RoleName, *role.CreateDate)
+				if err = detachPolicies(client, role.RoleName); err != nil {
+					return err
 				}
-			}
-			expirationDate := time.Now().UTC().Add(keepDuration)
-			if hasPrefix && expirationDate.After(*role.CreateDate) && (role.RoleLastUsed == nil || expirationDate.After(*role.RoleLastUsed.LastUsedDate)) {
-				var policiesMarker *string
-				for {
-					larpi := &iam.ListAttachedRolePoliciesInput{RoleName: role.RoleName, Marker: policiesMarker}
-					larpo, err := client.ListAttachedRolePolicies(larpi)
-
-					if err != nil {
-						log.Fatalf("Failed to get policies for %s because of %v", *role.RoleName, err)
-					}
-
-					for _, policy := range larpo.AttachedPolicies {
-						drpi := &iam.DetachRolePolicyInput{PolicyArn: policy.PolicyArn, RoleName: role.RoleName}
-						_, err = client.DetachRolePolicy(drpi)
-						if err != nil {
-							log.Fatalf("Failed to detach policy %s from %s because of %v", *policy.PolicyName, *role.RoleName, err)
-						}
-					}
-
-					if larpo.Marker == nil {
-						break
-					}
-					policiesMarker = larpo.Marker
+				if err = deleteProfiles(client, expirationDate, role.RoleName); err != nil {
```

Comment:
I thought we would need to delete Profiles, then Roles, then Policies instead of this order though based on this error from the workflow https://github.com/aws-observability/aws-otel-collector/runs/5008125297?check_suite_focus=true

Reply:
That's the order I'm doing it in. I'm detaching the policies, deleting the profiles, deleting the roles, and then deleting the policies.